content/blog/2022-04-07-compiling-black-with-mypyc-part-1.md

```diff
@@ -0,0 +1,591 @@
+
+             The calls increasingly `omit` right-hand trailers (bracket pairs with
+@@ -362,6 +365,11 @@ def transform_line(
+                 line, line_length=mode.line_length, features=features
+             )
+
++        # HACK: functions (like rhs) compiled by mypyc don't retain their __name__
++        # attribute which is needed in `run_transformer` further down. Unfortunately
++        # a nested class breaks mypyc too. So a class must be created via type ...
++        rhs = type("rhs", (), {"__call__": _rhs})()
++
+         if mode.experimental_string_processing:
+             if line.inside_brackets:
+                 transformers = [
+```
+
+I also had to change the code accessing `__name__` to do so via `__class__` so this hack
+could work. Minor correction though, top-level compiled functions still have `__name__`,
+it's just nested functions that don't have 'em[^8]. I improved this comment before this
+landed.
+
+#### The least bad workaround
+
+I remember getting an impromptu "how to read mypyc-generated C code" lesson from
+[@JelleZijlstra] [^9]. I was trying to debug this runtime crash in this `black.trans`
+class:
+
+```python3 {hl_lines=[6]}
+class StringParser:
+    """
+    [snipped]
+    """
+
+    DEFAULT_TOKEN = -1
+
+    # String Parser States
+    START = 1
+    DOT = 2
+    NAME = 3
+    PERCENT = 4
+    SINGLE_FMT_ARG = 5
+    LPAR = 6
+    RPAR = 7
+    DONE = 8
+
+    # Lookup Table for Next State
+    _goto: Dict[Tuple[ParserState, NodeType], ParserState] = {
+        [snipped]
+    }
+
+    [snipped]
+```
+
+```python3
+Traceback (most recent call last):
+  File "<stdin>", line 1, in <module>
+  File "test2.py", line 30, in <module>
+    (START, DEFAULT_TOKEN): DONE,
+KeyError: 'DEFAULT_TOKEN
+```
+
+It turns out the generated code wanted to access the `DEFAULT_TOKEN` attribute from the
+globals dict, predictably blowing up.
+
+```c {hl_lines=[1]}
+    cpy_r_r146 = CPyStatic_globals;
+    cpy_r_r147 = CPyStatics[17]; /* 'DEFAULT_TOKEN' */
+    cpy_r_r148 = CPyDict_GetItem(cpy_r_r146, cpy_r_r147);
+```
+
+If this scares you, don't fret, it's not too hard to read. First, the globals dict is
+assigned to `cpy_r_r146`, then the name (or key) we're looking up is read and stored in
+`cpy_r_r147`. Finally, `cpy_r_r146` and `cpy_r_r147` are passed to the `CPyDict_GetItem`
+function.
+
+Effectively, the buggy C code was trying to do this (unnecessary lines were replaced with
+`...`):
+
+```python
+class StringParser:
+    DEFAULT_TOKEN = -1
+    START = 1
+    DONE = 8
+    ...
+
+    __var_146 = globals()
+    __var_147 = "DEFAULT_TOKEN"
+    __var_148 = var_146[var_147]
+
+    _goto = {
+        (START, __var_148): DONE
+        ...
+    }
+```
+
+The root issue ([mypyc#862]) is that mypyc doesn't understand class level scoping ... or
+that negative integers are constants (using a positive integer fixed the crash). Working
+around this issue allowed me to embed the date I fixed this and many other issues. I think
```

Comment:
```suggestion
around this issue allowed me to embed the date. I fixed this and many other issues. I think
```
Not sure if a period is the right thing here, but it seems to be missing some form of punctuation.

Reply:
I'm just going to rewrite this sentence, it's poorly constructed anyway.